src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp

```diff
@@ -269,27 +276,20 @@ void MergeTreeBackgroundExecutor<Queue>::routine(TaskRuntimeDataPtr item)
 
     String query_id;
 
-    auto complete_task = [this, &erase_from_active, &release_task, &query_id] (TaskRuntimeDataPtr && item_)
+    auto complete_task = [this, &erase_from_active, &release_task] (TaskRuntimeDataPtr && item_)
```

Comment:
This comment about `release_task`. I could not attach it to that piece of code.

As I understand, `auto restart_task` still could do cancel under the lock. Sould it also be replaced out of mutex scope?

Reply:
Yes, there's still a problem with `restart_task`, will fix it in a follow-up PR. Btw, can we call `cancel for them in `removeTasksCorrespondingToStorage` after waiting for these tasks? We would not need the `is_currently_deleting` flag